File: examples/typescript-vitest/steps/library.steps.ts
import { defineState } from '@oselvar/var'
import { FEE_PENCE_PER_DAY, type Loan, lateFee, mayBorrow } from './library'

const MONTHS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
] as const

// Custom parameter types are declared inline so their transformer return types
// flow into the steps: {date} → Date, {money} → pence, {title} → string. The
// step handlers below need no argument annotations as a result.
const { stimulus, sensor } = defineState(
  () => ({ loans: [] as ReadonlyArray<Loan>, feePence: 0, granted: false }),
  {
    date: {
      // June 6th → the ISO date 2026-06-06 (the spec's year is 2026)
      regexp:
        /(?:January|February|March|April|May|June|July|August|September|October|November|December) \d{1,2}(?:st|nd|rd|th)/,
      transformer: (raw: string) => {
        const [month = '', day = ''] = raw.split(' ')
        const mm = String(MONTHS.indexOf(month as (typeof MONTHS)[number]) + 1).padStart(2, '0')
        const dd = String(Number.parseInt(day, 10)).padStart(2, '0')
        return `2026-${mm}-${dd}`
      },
    },
    money: {
      // £2.50 and 50p, both as pence
      regexp: /£\d+(?:\.\d{2})?|\d+p/,
      transformer: (raw: string) =>
        raw.startsWith('£') ? Math.round(Number(raw.slice(1)) * 100) : Number.parseInt(raw, 10),
    },
    title: {
      // Emphasis (*Emma*) is stripped before matching, so a title is a
      // Title Case run in the plain prose
      regexp: /[A-Z][a-z]+(?: [A-Z][a-z]+)*/,
      transformer: (raw: string) => raw,
    },
  },
)

stimulus('borrowed {title}, due back on {date}', (state, title, due) => ({
  loans: [...state.loans, { title, due }],
}))

stimulus('returns it on {date}', (state, returnedOn) => ({
  feePence: state.loans.reduce((fee, loan) => fee + lateFee(loan, returnedOn), 0),
}))

sensor('owes a {money} late fee', (state) => state.feePence)

sensor('{money} for each day overdue', () => FEE_PENCE_PER_DAY)

stimulus('asks to borrow {title} on {date}', (state, _title, on) => ({
  granted: mayBorrow(state.loans, on),
}))

sensor('the library refuses', (state) => {
  if (state.granted) throw new Error('expected the library to refuse')
})

sensor('the library agrees', (state) => {
  if (!state.granted) throw new Error('expected the library to agree')
})

File: examples/go-gotest/library.steps.go
package example

import (
	"errors"
	"strings"

	"github.com/varar-dev/varar-go/varar"
)

func registerLibrary(s *varar.Steps[Ctx]) {
	s.Param("date", `[A-Z][a-z]+ \d{1,2}, \d{4}`,
		func(g []string) varar.Value { return ParseDate(g[0]).EncodeVarValue() },
		func(v varar.Value) (string, bool) {
			var d Date
			if err := d.DecodeVarValue(v); err != nil {
				return "", false
			}
			return FormatDate(d), true
		})

	s.Param("money", `£\d+(?:\.\d+)?|\d+p`,
		func(g []string) varar.Value { return varar.IntValue(ParseMoney(g[0])) },
		func(v varar.Value) (string, bool) {
			if p, ok := v.AsInt(); ok {
				return FormatMoney(p), true
			}
			return "", false
		})

	s.Param("title", `\*[^*]+\*`,
		func(g []string) varar.Value {
			return varar.StrValue(strings.TrimSuffix(strings.TrimPrefix(g[0], "*"), "*"))
		},
		func(v varar.Value) (string, bool) {
			if t, ok := v.AsString(); ok {
				return "*" + t + "*", true
			}
			return "", false
		})

	s.Stimulus("borrowed {title}, due back on {date}", func(ctx Ctx, title string, due Date) (Ctx, error) {
		ctx.Loans = append(append([]Loan{}, ctx.Loans...), Loan{Title: title, Due: due})
		return ctx, nil
	})

	s.Stimulus("returns it on {date}", func(ctx Ctx, returned Date) (Ctx, error) {
		fee := 0
		for _, loan := range ctx.Loans {
			fee += int(LateFee(loan.Due, returned))
		}
		ctx.Fee = fee
		return ctx, nil
	})

	s.Sensor("owes a {money} late fee", func(ctx Ctx, expected int) (int, error) {
		return ctx.Fee, nil
	})

	s.Sensor("{money} for each day overdue", func(ctx Ctx, expected int) (int, error) {
		return int(FeePerDay), nil
	})

	s.Stimulus("asks to borrow {title} on {date}", func(ctx Ctx, title string, on Date) (Ctx, error) {
		dues := make([]Date, len(ctx.Loans))
		for i, loan := range ctx.Loans {
			dues[i] = loan.Due
		}
		ctx.Granted = MayBorrow(dues, on)
		return ctx, nil
	})

	s.Sensor("the library refuses", func(ctx Ctx) error {
		if ctx.Granted {
			return errors.New("expected the library to refuse")
		}
		return nil
	})

	s.Sensor("the library agrees", func(ctx Ctx) error {
		if !ctx.Granted {
			return errors.New("expected the library to agree")
		}
		return nil
	})
}
